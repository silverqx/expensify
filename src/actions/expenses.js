import database from '../firebase/firebase'
import * as types from '../constants/expenseActionTypes'

export const addExpense = (expense) => ({
    type: types.ADD_EXPENSE,
    expense
})

export const startAddExpense = ({
    description = '',
    note = '',
    amount = 0,
    createdAt = 0
} = {}) => {
    return (dispatch) => {
        const expense = {
            description,
            note,
            amount,
            createdAt
        }

        return database.ref('expenses').push(expense).then((snapshot) => {
            dispatch(addExpense({
                id: snapshot.key,
                ...expense
            }))
        })
    }
}

export const editExpense = (id, updates) => ({
    type: types.EDIT_EXPENSE,
    id,
    updates
})

export const removeExpense = (id) => ({
    type: types.REMOVE_EXPENSE,
    id
})
