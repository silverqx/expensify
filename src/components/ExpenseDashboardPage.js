import React from 'react'

import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
import ExpensesSummary from './ExpensesSummary'

const ExpenseDashboardPage = () => (
    <div>
        <h1>Expense Dashboard Page</h1>
        <h3>Expenses List</h3>
        <ExpensesSummary />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
)

export default ExpenseDashboardPage
