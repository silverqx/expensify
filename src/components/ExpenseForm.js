import React, { Component } from 'react'

import moment from 'moment'
// TODO check this along with import 'react-dates/lib/css/_datepicker.css' in App.js silver
import 'react-dates/initialize'
import { SingleDatePicker } from 'react-dates'
import debounce from 'lodash.debounce'

import FormErrors from './FormErrors'

export default class ExpenseForm extends Component {
    constructor(props) {
        super(props)

        const expense = props.expense || {}

        this.state = {
            expense: {
                description: expense.description || '',
                amount: expense.amount ? expense.amount + '' : '',
                createdAt: expense.createdAt ? moment(expense.createdAt) : moment(),
                note: expense.note || ''
            },
            datepickerFocused: false,
            errors: []
        }
    }

    componentWillUnmount() {
        this.validateFormDebounced.cancel()
    }

    onDescriptionChange = (e) => {
        const description = e.target.value

        this.setState((prevState) => ({
            expense: {
                ...prevState.expense,
                description
            }
        }))

        this.validateFormDebounced()
    }

    onAmountChange = (e) => {
        const amount = e.target.value

        // TODO take comma into account, eg try sk locale silver
        if (amount && !amount.match(/^\d+((\.|\,)\d{0,2})?$/))
            return

        this.setState((prevState) => ({
            expense: {
                ...prevState.expense,
                amount
            }
        }))

        this.validateFormDebounced()
    }

    onDateChange = (createdAt) => {
        this.setState((prevState) => ({
            expense: {
                ...prevState.expense,
                createdAt
            }
        }))
    }

    onNoteChange = (e) => {
        const note = e.target.value

        this.setState((prevState) => ({
            expense: {
                ...prevState.expense,
                note
            }
        }))
    }

    validateFormDebounced = debounce(
        () => this.validateForm(), 600
    )

    validateForm = () => {
        const expense = this.state.expense
        let errors = []

        if (!expense.description)
            errors = errors.concat('Description is required.')
        if (!expense.amount)
            errors = errors.concat('Amount is required.')
        if (!expense.createdAt)
            errors = errors.concat('CreatedAt is required.')

        this.setState(() => ({ errors }))

        return !!!errors.length
    }

    onSubmit = (e) => {
        e.preventDefault()

        if (!this.validateForm())
            return

        const expense = this.state.expense

        this.props.onSubmit({
            ...expense,
            amount: parseFloat(expense.amount, 10) || 0,
            createdAt: expense.createdAt.valueOf()
        })
    }

    render() {
        return (
            <div>
                <FormErrors errors={this.state.errors} />
                <form onSubmit={this.onSubmit}>
                    <div>
                        <input
                            type="text"
                            autoFocus
                            maxLength={255}
                            placeholder="Description..."
                            value={this.state.expense.description}
                            onChange={this.onDescriptionChange}
                        />
                    </div>
                    <div>
                        <input
                            type="number"
                            placeholder="Amount..."
                            step={1.0}
                            value={this.state.expense.amount}
                            onChange={this.onAmountChange}
                            step={0.5}
                        />
                    </div>
                    <div>
                        <SingleDatePicker
                            id="created-at"
                            date={this.state.expense.createdAt}
                            onDateChange={this.onDateChange}
                            focused={this.state.datepickerFocused}
                            onFocusChange={({ focused }) => this.setState({ datepickerFocused: focused })}
                            isOutsideRange={() => false}
                            numberOfMonths={1}
                            placeholder="Created At..."
                            small={true}
                            daySize={32}
                        />
                    </div>
                    <div>
                        <textarea
                            cols="30"
                            rows="4"
                            placeholder="Note... (optional)"
                            value={this.state.expense.note}
                            onChange={this.onNoteChange}
                        ></textarea>
                    </div>
                    <div>
                        <input
                            type="submit"
                            value={this.props.expense ? 'Edit Expense' : 'Add Expense'}
                        />
                    </div>
                </form>
            </div>
        )
    }
}
