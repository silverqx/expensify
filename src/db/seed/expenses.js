import moment from 'moment'

const expensesSeed = [{
    description: 'Water Bill',
    note: 'Water Bill Note',
    amount: 166.60,
    createdAt: moment('2019-04-19').valueOf()
}, {
    description: 'Gas Bill',
    note: 'Gas Bill Note',
    amount: 5.20,
    createdAt: moment('2019-03-17').valueOf()
}, {
    description: 'Internet Bill',
    note: 'Internet Bill Note',
    amount: 50.20,
    createdAt: moment('2019-04-09').valueOf()
}, {
    description: 'Telekom Bill',
    note: 'Telekom Bill Note',
    amount: 10.20,
    createdAt: moment('2019-01-10').valueOf()
}]

export default expensesSeed
