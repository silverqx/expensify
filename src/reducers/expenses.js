import * as types from '../constants/expenseActionTypes'

const expensesReducerInitialState = []

const expensesReducer = (state = expensesReducerInitialState, action) => {
    switch (action.type) {
        case types.ADD_EXPENSE:
            return [
                ...state,
                action.expense
            ]
        case types.EDIT_EXPENSE:
            return state.map((expense) => {
                if (expense.id !== action.id)
                    return expense

                return {
                    ...expense,
                    ...action.updates
                }
            })
        case types.REMOVE_EXPENSE:
            return state.filter((expense) => expense.id !== action.id)
        case types.SET_EXPENSES:
            return action.expenses
        default:
            return state
    }
}

export default expensesReducer
