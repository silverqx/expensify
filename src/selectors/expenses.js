import moment from 'moment'

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())

        const createdAt = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAt, 'day') : true
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAt, 'day') : true

        return textMatch && startDateMatch && endDateMatch
    }).sort((a, b) => {
        const sortByMap = {
            date: 'createdAt',
            amount: 'amount'
        }
        const sortColumn = sortByMap[sortBy]

        if (a[sortColumn] < b[sortColumn])
            return -1
        if (a[sortColumn] > b[sortColumn])
            return 1

        return 0
    })
}

export default getVisibleExpenses
