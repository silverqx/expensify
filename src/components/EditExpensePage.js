import React, { Component } from 'react'
import { connect } from 'react-redux'

import ExpenseForm from './ExpenseForm'

import { startEditExpense, startRemoveExpense } from '../actions/expenses'

export class EditExpensePage extends Component {
    onSubmit = (updatedExpense) => {
        this.props.startEditExpense(
            this.props.expense.id,
            updatedExpense
        )
        this.props.history.push('/')
    }

    onRemove = () => {
        this.props.startRemoveExpense(
            this.props.expense.id
        )
        this.props.history.push('/')
    }

    onCancel = () => {
        this.props.history.push('/')
    }

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm
                        expense={this.props.expense}
                        onSubmit={this.onSubmit}
                        onRemove={this.onRemove}
                        onCancel={this.onCancel}
                    />
                </div>
            </div>
        )
    }
}

// FIXME really bad solution, should be done in router? And how to test this? silver
const findExpense = (state, { history, match }) => {
    const expense = state.expenses.find(
        expense => expense.id === match.params.id
    )

    if (!expense)
        return history.goBack()

    return expense
}

const mapStateToProps = (state, props) => ({
    expense: findExpense(state, props)
})

const mapDispatchToProps = (dispatch) => ({
    startEditExpense: (id, updatedExpense) => dispatch(
        startEditExpense(id, updatedExpense)
    ),
    startRemoveExpense: id => dispatch(startRemoveExpense(id)),
})

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(EditExpensePage)
