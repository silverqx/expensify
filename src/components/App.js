import React, { Component } from 'react'
import { Router, Route, Switch } from "react-router-dom"
import ReactModal from 'react-modal'
import { createBrowserHistory } from 'history'

import ExpenseDashboardPage from './ExpenseDashboardPage'
import AddExpensePage from './AddExpensePage'
import EditExpensePage from './EditExpensePage'
import LoginPage from './LoginPage'
import NotFoundPage from './NotFoundPage'

import ModalOption from './ModalOption'
import PrivateRoute from '../routers/PrivateRoute'
import PublicRoute from '../routers/PublicRoute'

// import { clearInputValue } from '../utils/utils'

import 'react-dates/initialize'
import 'react-dates/lib/css/_datepicker.css'

// Make sure to bind modal to your appElement
// (http://reactcommunity.org/react-modal/accessibility/)
ReactModal.setAppElement('#app')

export const history = createBrowserHistory()

export default class ExpensifyApp extends Component {
    static defaultProps = {
        title: 'Expensify'
    }

    state = {
        error: null,
        // TODO implement subtitle silver
        subtitle: 'Xxx',
    }

    componentDidMount() {
    }

    componentDidUpdate(prevProps, prevState) {
    }

    /**
     * Update Error state.
     *
     * @param {string} error - Error message.
     *
     * @returns {?boolean} True every time.
     */
    handleError = (error) => {
        this.setState(() => ({ error }))

        return true
    }

    /**
     * Clear Error state.
     *
     * @returns {?boolean} False every time.
     */
    clearError = () => {
        this.setState(() => ({ error: null}))

        return false
    }

    render() {
        return (
            <Router history={history}>
                <Switch>
                    <PublicRoute path="/" exact component={LoginPage} />
                    <PrivateRoute path="/dashboard" component={ExpenseDashboardPage} />
                    <PrivateRoute path="/create" component={AddExpensePage} />
                    <PrivateRoute path="/edit/:id" component={EditExpensePage} />
                    <Route component={NotFoundPage} />
                </Switch>
                <ModalOption
                    selectedOption={this.state.selectedOption}
                    handleClearSelectedOption={this.handleClearSelectedOption}
                />
            </Router>
        )
    }
}
