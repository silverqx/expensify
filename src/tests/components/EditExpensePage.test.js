import React from 'react'
import { shallow } from 'enzyme'

import { EditExpensePage } from '../../components/EditExpensePage'

import { expensesWithIds as expenses } from '../fixtures/expenses'

describe('EditExpensePage component', () => {
    let startEditExpense,
        startRemoveExpense,
        history,
        wrapper

    beforeEach(() => {
        startEditExpense = jest.fn()
        startRemoveExpense = jest.fn()
        history = { push: jest.fn() }

        wrapper = shallow(
            <EditExpensePage
                expense={expenses[0]}
                startEditExpense={startEditExpense}
                startRemoveExpense={startRemoveExpense}
                history={history}
            />
        )
    })

    test('should render EditExpensePage correctly', () => {
        expect(wrapper).toMatchSnapshot()
    })

    test('should handle startEditExpense on form onSubmit', () => {
        const updatedExpense = {
            ...expenses[0],
            description: 'Water Bill Updated',
            amount: 155.25
        }

        wrapper.find('ExpenseForm').prop('onSubmit')(updatedExpense)

        expect(startEditExpense).toHaveBeenCalledTimes(1)
        expect(startEditExpense).toHaveBeenCalledWith(
            expenses[0].id,
            updatedExpense
        )

        expect(history.push).toHaveBeenCalledTimes(1)
        expect(history.push).toHaveBeenCalledWith('/')
    })

    test('should go to home on form onCancel', () => {
        wrapper.find('ExpenseForm').prop('onCancel')()

        expect(history.push).toHaveBeenCalledTimes(1)
        expect(history.push).toHaveBeenCalledWith('/')
    })

    test('should handle startRemoveExpense on form onRemove', () => {
        wrapper.find('ExpenseForm').prop('onRemove')()

        expect(startRemoveExpense).toHaveBeenCalledTimes(1)
        // 1 is expense id
        expect(startRemoveExpense).toHaveBeenCalledWith('1')

        expect(history.push).toHaveBeenCalledTimes(1)
        expect(history.push).toHaveBeenCalledWith('/')
    })
})
