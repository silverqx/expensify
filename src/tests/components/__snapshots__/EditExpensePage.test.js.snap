// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditExpensePage component should render EditExpensePage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 166.6,
          "createdAt": 1555624800000,
          "description": "Water Bill",
          "id": "1",
          "note": "Water Bill Note",
        }
      }
      onCancel={[Function]}
      onRemove={[Function]}
      onSubmit={[Function]}
    />
  </div>
</div>
`;
